fixed percentage behaviour in instr

renatoviana12 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind api-change
/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes  https://github.com/kubernetes/kubernetes/issues/89082

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

renatoviana12 committed:
Fixed percentage behavior in instr

fixed syntax, wrote a test

fixed a test

.

1

Update staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

added test

.

fix

fix test

fixed a test

gofmt

lint

fix

function name

validation fix

.

godocs added

.

mikedanese commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -179,7 +179,11 @@ func getIntOrPercentValue(intOrStr *IntOrString) (int, bool, error) {
+			return int(v), true, nil
+		} else {
+			return int(v), false, nil
+		}
```
This is the same as
```
  return int(v), strings.Contains(intOrStr.StrVal, "%"), nil
```
Or
```
  return int(v), s != intOrStr.StrVal, nil
```

renatoviana12 replied:
You're right, I'll make some changes.

mikedanese commented:
Please add a test and sign the CLA.

Side note: it looks like "1%0" is treated the same as 10% (???). Should we fix that?

renatoviana12 commented:
> Please add a test and sign the CLA.
> 
> Side note: it looks like "1%0" is treated the same as 10% (???). Should we fix that?

This behaviour also doesn't seem correct but I wasn't sure if this PR was the correct place to do this change. With the current behaviour "1%0" is parsed as 10%. This is easy to correct, let me know if I should do it on this branch.

fedebongio commented:
/assign @jpbetz

jpbetz commented:
> > Please add a test and sign the CLA.
> > Side note: it looks like "1%0" is treated the same as 10% (???). Should we fix that?
> 
> This behaviour also doesn't seem correct but I wasn't sure if this PR was the correct place to do this change. With the current behaviour "1%0" is parsed as 10%. This is easy to correct, let me know if I should do it on this branch.

I'm in favor of correcting it. It would be pretty bad to depend on that behavior, so I'm not particularly worried about breaking anyone by fixing it.

tanjunchen commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -109,6 +109,15 @@ func TestIntOrStringMarshalJSONUnmarshalYAML(t *testing.T) {
+	intorstr.StrVal = "1"
+
+	_, perc, _ := getIntOrPercentValue(intorstr)
+	if perc == true {
```
```suggestion
	if perc {
```

tanjunchen commented:
Please sign the CLA.
@renatoviana12 
any update?

renatoviana12 commented:
I have signed and have my github email properly configured, it's still not passing

tanjunchen commented:
/check-cla

tanjunchen commented:
/check-cla
/retest

renatoviana12 commented:
my bad, I'll rewrite the test asap

mikedanese commented:
@renatoviana12 the email on this patch:

https://github.com/kubernetes/kubernetes/commit/0c3487f939165b18a7fc5d9b50a9382eca541013.patch

Is not an email configured with your github account. Is that expected? You can change the address on the commit with `git commit --amend --author "Renato Viana <different@email.com>"`.

renatoviana12 commented:
/check-cla
/retest

tanjunchen commented:
/ok-to-test
/check-cla
/retest

tanjunchen commented:
/retest

renatoviana12 commented:
Is this ready to merge?

tanjunchen commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -179,7 +179,8 @@ func getIntOrPercentValue(intOrStr *IntOrString) (int, bool, error) {
 			return 0, false, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
 		}
-		return int(v), true, nil
+		return int(v), strings.Contains(intOrStr.StrVal, "%"), nil
```
see   https://github.com/kubernetes/kubernetes/pull/89482/files#diff-31ae0ffb0640d8ca5eb91a4a72e8f152R177

it seems that we have replace the '%'

tanjunchen replied:
```
case intstr.String:
		isPercent := false
		s := intOrStr.StrVal
		if strings.Contains(s, "%") {
			isPercent = true
			s = strings.Replace(intOrStr.StrVal, "%", "", -1)
		}
		v, err := strconv.Atoi(s)
		if err != nil {
			return 0, isPercent, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
		}
		return int(v), isPercent, nil
	}
```

tanjunchen replied:
/cc @JoelSpeed

JoelSpeed replied:
https://github.com/kubernetes/kubernetes/pull/89482#discussion_r411065116 This suggestion seems good to me, though I would suggest swapping `strings.Contains` for `strings.HasSuffix` and `strings.Replace` for `strings.TrimSuffix` so that cases where a user has entered `1%0` throws an error rather than is interpreted as `10%`

tanjunchen replied:
agree. could you think about it? it looks good.

tanjunchen replied:
swapping strings.Contains for strings.HasSuffix and strings.Replace for strings.TrimSuffix . 
And add some testcase.

renatoviana12 replied:
Looks good to me. I'll make the changes asap

renatoviana12 replied:
commited the changes

renatoviana12 replied:
@tanjunchen @JoelSpeed did you agree with the implementation?

JoelSpeed replied:
Implementation looks good, I added a suggestion for an extra test case, you may also be inspired for some more test cases by https://github.com/openshift/machine-api-operator/blob/b9f8d296e5f713a8fe7c5b89013ab37a4ac69172/pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go#L2645-L2719

renatoviana12 replied:
Thanks, added some more test cases, commiting the changes now

renatoviana12 commented:
Fixed the failing test

JoelSpeed commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -110,6 +110,61 @@ func TestIntOrStringMarshalJSONUnmarshalYAML(t *testing.T) {
+			expectPerc: false,
+		},
+		{
+			input:      FromString("40%"),
```
Would be good to include a test case for `FromString("200")` as well to prove `expectPerc` would be false

renatoviana12 replied:
Commited the new tests, do they look good?
@tanjunchen

JoelSpeed commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -110,6 +110,87 @@ func TestIntOrStringMarshalJSONUnmarshalYAML(t *testing.T) {
+			expectPerc: false,
+		},
+	}
+	for i, test := range tests {
+		t.Logf("test case %d", i)
+		value, isPercent, err := getIntOrPercentValue(&test.input)
+		if test.expectVal != value {
+			t.Errorf("expected value does not match")
+			continue
+		}
+		if test.expectPerc && isPercent == false {
+			t.Errorf("expected percentage, but got not percentage")
+			continue
+		}
+		if !test.expectPerc && isPercent == true {
+			t.Errorf("expected not percentage and got percentage")
+			continue
+		}
+		if test.expectErr && err == nil {
+			t.Errorf("expected error, but got none")
+			continue
+		}
+		if !test.expectErr && err != nil {
+			t.Errorf("unexpected err: %v", err)
+			continue
+		}
+		if test.expectVal != value {
+			t.Errorf("expected %v, but got %v", test.expectVal, value)
+		}
+	}
```
This could be condensed a lot, I'd also recommend using `t.Run` so execute a subtest here, then instead of Error and Continue you can Fatal the subtest
```suggestion
	for _, test := range tests {
	    t.Run("", func(t *testing.T) {
	        value, isPercent, err := getIntOrPercentValue(&test.input)
		if test.expectVal != value {
			t.Fatalf("expected value does not match, expected: %d, got: %d", test.expectVal, value)
		}
		if test.expectPerc != isPercent {
			t.Fatalf("expected percent does not match, expected: %d, got: %d", test.expectPerc, isPercent)
		}
		if test.expectErr != (err != nil) {
			t.Fatalf("expected error does not match, expected error: %v, got: %v", test.expectErr, err)
		}
	    })	
	}
```

renatoviana12 replied:
Looks good to me, commited the suggestion

JoelSpeed commented:
Assuming tests all pass, I'm happy with the code now

/lgtm

enxebre commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -110,6 +110,74 @@ func TestIntOrStringMarshalJSONUnmarshalYAML(t *testing.T) {
+			expectPerc: false,
+		},
+		{
+			input:      FromString("30%0"),
```
would also be good to include "40%%" here

renatoviana12 replied:
added the test

renatoviana12 commented:
@JoelSpeed I added another test case as @enxebre suggested. How does it look now?

JoelSpeed reviewed (commented):
/lgtm

Good catch @enxebre, thanks for fixing @renatoviana12

mikedanese commented:
/lgtm
/approve

renatoviana12 commented:
/assign @deads2k

renatoviana12 commented:
Do I need to do something else to get this ready to merge?

renatoviana12 commented:
Does this still make sense to merge?

JoelSpeed commented:
I believe the bug is still present so I would say so, do you have time to rebase? Note we are in code freeze right now so it won't be able to merge until thaw which should happen in a couple of weeks time

renatoviana12 commented:
Yeah, I'll get to it

renatoviana12 commented:
@JoelSpeed rebase done :)

JoelSpeed reviewed (approved):
/lgtm

renatoviana12 commented:
/remove-lifecycle rotten

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -174,12 +174,17 @@ func getIntOrPercentValue(intOrStr *IntOrString) (int, bool, error) {
 	case Int:
 		return intOrStr.IntValue(), false, nil
 	case String:
```
reviewing call sites for safety:

```
getIntOrPercentValue
	GetValueFromIntOrPercent
		DaemonSetsController#getUnavailableNumbers
		NewRSNewReplicas(deployment)
		ResolveFenceposts
			MaxUnavailable(deployment apps.Deployment)
			MaxSurge(deployment apps.Deployment)
		DisruptionController#getExpectedPodCount(pdb)
	ValidateRollingUpdateDeployment
	ValidateRollingUpdateDaemonSet
```

called on these API fields:
* `DaemonSet.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable`

    String values are guaranteed to match `^[0-9]+%$` by `ValidatePositiveIntOrPercent`: https://github.com/kubernetes/kubernetes/blob/ab3ed8c19cf3fde87407ccacbc68939421585346/pkg/apis/apps/validation/validation.go#L344-L347

* `Deployment.Spec.Strategy.RollingUpdate.MaxSurge`
* `Deployment.Spec.Strategy.RollingUpdate.MaxUnavailable`

    String values are guaranteed to match `^[0-9]+%$` by `ValidatePositiveIntOrPercent`: https://github.com/kubernetes/kubernetes/blob/ab3ed8c19cf3fde87407ccacbc68939421585346/pkg/apis/apps/validation/validation.go#L432-L436

* `PodDisruptionBudget.Spec.MaxUnavailable`
* `PodDisruptionBudget.Spec.MaxAvailable`

    String values are guaranteed to match `^[0-9]+%$` by `ValidatePositiveIntOrPercent`: https://github.com/kubernetes/kubernetes/blob/ab3ed8c19cf3fde87407ccacbc68939421585346/pkg/apis/policy/validation/validation.go#L47-L64

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -174,12 +174,17 @@ func getIntOrPercentValue(intOrStr *IntOrString) (int, bool, error) {
 		v, err := strconv.Atoi(s)
 		if err != nil {
-			return 0, false, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
+			return 0, isPercent, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
```
```suggestion
			return 0, false, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
```

return values other than the error are not intended to be meaningful

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -174,12 +174,17 @@ func getIntOrPercentValue(intOrStr *IntOrString) (int, bool, error) {
+			return 0, isPercent, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
 		}
-		return int(v), true, nil
+		return int(v), isPercent, nil
```
Returning `isPercent ` here is a safe change for in-tree uses, since all the string values we pass to this method were already well-formatted percentages.

However, this is a breaking change for callers that passed in numeric strings (e.g. "10"). Previously, that would return `10, true, nil`. Now it returns `10, false, nil`.

Searching usages (https://grep.app/search?q=.GetValueFromIntOrPercent%28&filter[lang][0]=Go) this method is called externally.

If someone had documented their API to mean "integers are treated as absolute values, strings are treated as percentages", this change will break them.

liggitt reviewed (commented):
cc @deads2k @lavalamp on the compatibility question for this fix

lavalamp commented:
I suggest:
* Leave the existing function alone, add "deprecated" to the comment.
* Add a parallel function that treats ints as ints, strings that are a number followed by a single percent sign as a percent, and all other strings as an error
* Fix-up in-tree references to call the new function.

renatoviana12 commented:
> I suggest:
> 
> * Leave the existing function alone, add "deprecated" to the comment.
> * Add a parallel function that treats ints as ints, strings that are a number followed by a single percent sign as a percent, and all other strings as an error
> * Fix-up in-tree references to call the new function.


This seems like a sensible solution. Should I go forward with this?

liggitt commented:
> This seems like a sensible solution. Should I go forward with this?

sounds good to me

renatoviana12 commented:
/assign @liggitt
Deprecated the old function and created a new one called getIntOrPercentValueFromIntOrString that now replaces all other in-tree references.

renatoviana12 commented:
Added a linter exception on the method not being called in-tree

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -155,7 +155,7 @@ func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool
 		return 0, errors.New("nil value for IntOrString")
 	}
-	value, isPercent, err := getIntOrPercentValue(intOrPercent)
+	value, isPercent, err := getIntOrPercentValueFromIntOrString(intOrPercent)
```
`GetValueFromIntOrPercent` (the exported method that has out-of-tree callers) should still call getIntOrPercentValue and should be marked deprecated in favor of a corrected exported method.

renatoviana12 replied:
Created a new parallel function and added test cases extending the already existing ones to it. Also deprecated the old method.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -169,6 +169,8 @@ func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool
 }
 
+//Deprecated: use getIntOrPercentValueFromIntOrString
+//lint:ignore U1000 we want to keep the changes from breaking external uses of the method
```
`GetValueFromIntOrPercent` should still call this to avoid breaking external users, this lint exception shouldn't be required

renatoviana12 commented:
What is the appropriate way to deal with the linter acting up with the method being deprecated?

Adding a //lint:ignore SA1019 at every point the old method is called?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -151,6 +151,25 @@ func ValueOrDefault(intOrPercent *IntOrString, defaultValue IntOrString) *IntOrS
+}
+
+//Deprecated: use GetIntFromIntOrPercent
 func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool) (int, error) {
```
we want to limit the bad code to this public method... I would recommend the following:

1. inline the old `getIntOrPercentValue` definition into `GetValueFromIntOrPercent` (so we contain the bad behavior to a single method)
2. fix `getIntOrPercentValue` to have the behavior we want
3. add a new public method which calls to `getIntOrPercentValue` (a clear name will be tricky to choose), with godocs clearly explaining the behavior for int values, numeric string values, percent string values, and malformed string values
4. mark `GetValueFromIntOrPercent` as deprecated (as you've done) and make sure there are no in-tree calls to it
5. add godocs to `GetValueFromIntOrPercent` explaining in detail the problematic behavior and the reason to switch to the new public method

renatoviana12 replied:
Is GetIntFromIntOrPercent a good name for the new method?

liggitt replied:
not sure... `GetScaledValueFromIntOrPercent` might be a little clearer?

lavalamp replied:
GetValueFromIntOrPercentEx
GetValueFromIntOrPercentSafe
GetValueFromIntOrPercentWithErrorChecking

renatoviana12 replied:
Went with GetScaledValueFromIntOrPercent as that's effectively what the function does but as @liggitt said naming the new version of a function is tricky.

lavalamp replied:
"Scaled" is a very confusing name for this function. It isn't scaling anything.

renatoviana12 commented:
The old getIntOrPercentValue is now inlined into GetValueFromIntOrPercent. 
The new getIntOrPercentValue follows the behaviour discussed here. GetValueFromIntOrPercent is marked as deprecated and I added some explanation as to why. The new function is named GetScaledValueFromIntOrPercent as per @liggitt 's suggestion. Also added a godocs comment on GetScaledValueFromIntOrPercent. 

How do the changes look?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -151,7 +151,12 @@ func ValueOrDefault(intOrPercent *IntOrString, defaultValue IntOrString) *IntOrS
+//is a percentage string value it's treated as a percentage and scaled appropriately
+//in accordance to the total, if it's an int value it's treated as a a simple value and
+//if it is a numeric string value it returns an error.
+func GetScaledValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool) (int, error) {
```
You seem to have changed the name of the old function? You need to keep the old function the same name to not accidentally fix load bearing bugs.

liggitt replied:
that is git diffing unfortunately... `GetValueFromIntOrPercent` in this PR maintains the old code by inlining the old getIntOrPercentValue implementation as-is

liggitt replied:
(per my request at https://github.com/kubernetes/kubernetes/pull/89482#discussion_r488027200)

lavalamp replied:
oh, I see, the code looks the same because the function it called changed.

liggitt replied:
right, I wanted to contain the badness to the single deprecated exported method

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -169,17 +174,60 @@ func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool
+		return 0, fmt.Errorf("invalid value for IntOrString: %v", err)
+	}
+	if isPercent {
+		if roundUp {
```
This is kind of a horrifying block of code. There's actually (at least) four states people could want (floor, ceil, round, do nothing), and it's represented by a bool.

I recommend changing this function to return a float and let callers do the operation they need on this.

lavalamp replied:
(Just to be clear, I know you didn't write this!)

lavalamp replied:
This comment should be attached to the copy in the new function, of course :)

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -169,17 +174,60 @@ func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool
+//GetScaledValueFromIntOrPercent. This method was treating all int as a numeric value and all
+//strings with or without a percent symbol as a percentage value.
+func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool) (int, error) {
+	inlinedGetIntOrPercentValue := func(intOrStr *IntOrString) (int, bool, error) {
```
I don't think this needs to be an embedded function?

lavalamp replied:
I think it's less confusing to keep the existing non-embedded function, clearly comment not to call it, and make a new one with a different name for the new function.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -151,7 +151,12 @@ func ValueOrDefault(intOrPercent *IntOrString, defaultValue IntOrString) *IntOrS
+//is a percentage string value it's treated as a percentage and scaled appropriately
+//in accordance to the total, if it's an int value it's treated as a a simple value and
+//if it is a numeric string value it returns an error.
+func GetScaledValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool) (int, error) {
```
Can we delete the roundUp parameter and return a float?

Since we have to change call sites anyway, and the "rounding" done here is not good :)

lavalamp replied:
Wait, I think I misunderstood what this does given that "total" is passed in. I have to run and will look more later.

lavalamp replied:
OK I see now that I completely misunderstood what this function does, so sorry for the noise. Given that this is returning a specific number of things, the "scaled" name and rounding makes more sense.

The rounding still doesn't make a lot of sense abstractly, because actually choosing between floor and ceil is not the same as rounding. But I won't ask for a change, since it makes some sense if you're consuming discreet thingies.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -151,7 +151,12 @@ func ValueOrDefault(intOrPercent *IntOrString, defaultValue IntOrString) *IntOrS
 }
 
-func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool) (int, error) {
+//GetScaledValueFromIntOrPercent is meant to replace GetValueFromIntOrPercent.
```
Nit: please keep a space between the `//` and the words that follow to match the style of the code base.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -169,17 +174,60 @@ func GetValueFromIntOrPercent(intOrPercent *IntOrString, total int, roundUp bool
 	return value, nil
 }
 
+//Deprecated: GetValueFromIntOrPercent was deprecated in favor of
```
You'll break go lint if this comment doesn't start with the function name, so move the deprecation notice to be at the end of the comment, I guess.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -151,7 +151,12 @@ func ValueOrDefault(intOrPercent *IntOrString, defaultValue IntOrString) *IntOrS
+//This method returns a scaled value from an IntOrString type. If the IntOrString
+//is a percentage string value it's treated as a percentage and scaled appropriately
+//in accordance to the total, if it's an int value it's treated as a a simple value and
+//if it is a numeric string value it returns an error.
```
```suggestion
// if it is a string value which is either non-numeric or numeric but lacking a trailing '%' it returns an error.
```

renatoviana12 commented:
@lavalamp I updated the branch with your suggestions. I wish I could have taken care of this sooner but last week was hectic

lavalamp commented:
OK LGTM for the code change.

However the large number of call sites has me worried that we'll change behavior for existing incorrect-but-previously-acted-on configurations. I guess I'll defer to Jordan on that.

liggitt commented:
This LGTM to me. I swept the callsites in https://github.com/kubernetes/kubernetes/pull/89482#discussion_r482433865 and verified the fields being fed into this method were all validated using ValidatePositiveIntOrPercent, which ensures strings values match `^[0-9]+%$`, which means they are unaffected by the malformed string handling added here.

lavalamp commented:
/lgtm
/approve
